Framework: Add a QueryPurchases Component

Tug opened the pull request:
Part of https://github.com/Automattic/wp-calypso/issues/5046

This PR actually adds 2 new Query Components:
- `QuerySitePurchases` to query site purchases
- `QueryUserPurchases` to query user purchases

### Reviews

Those components aren't used anywhere yet but will be used in an upcoming PR to update components to use the new `state.purchases` reducer introduced in https://github.com/Automattic/wp-calypso/pull/6317 and https://github.com/Automattic/wp-calypso/pull/6383 

- [x] Code

@Automattic/sdev-feed 

Test live: https://calypso.live/?branch=add/query-purchases

gziolo commented on client/components/data/query-site-purchases/index.jsx:
```diff
@@ -0,0 +1,58 @@
+/**
+ * External dependencies
+ */
+import { Component, PropTypes } from 'react';
```
You need to provide `React` to make sure JSX transpilation works properly. It might work without, but I already learned myself that it's pretty random.

gziolo commented on client/components/data/query-user-purchases/index.jsx:
```diff
@@ -0,0 +1,58 @@
+/**
+ * External dependencies
+ */
+import { Component, PropTypes } from 'react';
```
Same as above. `React` needs to be always exposed explicitly when JSX is involved.

Tug replied:
You're right, but the thing is, `render` returns `null` in all Query Components and no JSX is ever used usually.
So I'm not sure this is really needed here.

gziolo replied:
I checked all existing Query components, and it turns out we never expose React. Interesting. :)
We should skip it here, too. So all good :)

gziolo commented on client/components/data/query-user-purchases/index.jsx:
```diff
@@ -0,0 +1,58 @@
+	fetchUserPurchases: PropTypes.func
+};
+
+QueryUserPurchases.defaultProps = {
```
Do we really need that?

gziolo commented on client/components/data/query-site-purchases/index.jsx:
```diff
@@ -0,0 +1,58 @@
+	fetchSitePurchases: PropTypes.func
+};
+
+QuerySitePurchases.defaultProps = {
```
We probably can skip that.

Tug replied:
Agreed, I'm not sure why they're defined in other Query Components

gziolo commented on client/components/data/query-site-purchases/index.jsx:
```diff
@@ -0,0 +1,58 @@
+}
+
+QuerySitePurchases.propTypes = {
+	siteId: PropTypes.number,
```
We should rather make all those required.

gziolo commented on client/components/data/query-user-purchases/index.jsx:
```diff
@@ -0,0 +1,58 @@
+	userId: PropTypes.number,
+	fetchingUserPurchases: PropTypes.bool,
+	fetchUserPurchases: PropTypes.func
+};
```
We should rather make all those required.

gziolo commented on client/components/data/query-user-purchases/index.jsx:
```diff
@@ -0,0 +1,58 @@
+	}
+
+	requestUserPurchases( props = this.props ) {
+		this.props.fetchUserPurchases( props.siteId );
```
Should be `userId`.

gziolo commented on client/components/data/query-user-purchases/index.jsx:
```diff
@@ -0,0 +1,58 @@
+
+	componentWillReceiveProps( nextProps ) {
+		if ( nextProps.fetchingUserPurchases ||
+			! nextProps.siteId ||
```
Should be `userId`.

gziolo commented:
Only copy and paste issues to resolve. Looks good otherwise ;)

gziolo commented on client/state/purchases/selectors.js:
```diff
@@ -5,3 +5,6 @@ import find from 'lodash/find';
 	assign( {}, state, { data: find( state.data, { id } ) } )
 );
+
+export const isFetchingUserPurchases = state => state.purchases.isFetchingUserPurchases;
```
We could add super simple unit tests for those method just to keep in sync with how other state selectors are covered with tests.

gziolo commented on client/components/data/query-site-purchases/index.jsx:
```diff
@@ -0,0 +1,54 @@
+}
+
+QuerySitePurchases.propTypes = {
+	siteId: PropTypes.number.required,
```
https://facebook.github.io/react/docs/reusable-components.html
`.isRequired` :)

gziolo replied:
It's hard to check if you can't run the code :D

Tug replied:
Haha yep, thanks!

Tug commented:
@gziolo Thanks for reviewing!
I have fixed those issues. Indeed if there are others we don't see we will when refactoring the purchases data components.

I have also added a 3rd commit to fix the format of selectors in `state/purchases`. If you don't think the fix is needed we can always remove it from this PR.

drewblaisdell commented on client/components/data/query-site-purchases/index.jsx:
```diff
@@ -0,0 +1,54 @@
+			fetchingSitePurchases: isFetchingSitePurchases( state )
+		};
+	},
+	{ fetchSitePurchases }
```
Do we need to bind this action to `dispatch`?

gziolo replied:
Yes, it's pretty cool :)

drewblaisdell commented:
👍

gziolo commented on client/state/purchases/test/reducer.js:
```diff
@@ -142,18 +142,11 @@ describe( 'items', () => {
-			isFetchingUserPurchases: false,
-			hasLoadedSitePurchasesFromServer: true,
-			hasLoadedUserPurchasesFromServer: true
+		expect( getByPurchaseId( { purchases: state }, 2 ) ).to.be.eql( {
```
We should extract now all tests related to `getByPurchaseId` to its own `selectors.js` test file.

Tug replied:
I think those are testing if the store updates for private registration cancellation. Granted those probably needs refactoring, but let's not change the logic of the tests in this PR ;)

gziolo replied:
I probably should check the full context, this is the limitation of GH diffs ;)
We probably can leave it as it is in that case.

gziolo commented on client/state/purchases/test/selectors.js:
```diff
@@ -2,31 +2,60 @@
+	} );
+
+	describe( 'isFetchingSitePurchases', () => {
+		it( 'should return a purchase by its ID, preserving the top-level flags', () => {
```
This description should be updated :)

gziolo commented:
Looks good, there are two more small tweaks that are needed before we merge, but those aren't blockers. I'm leaving the Ready to Merge label ;)

stephanethomas commented on client/components/data/query-site-purchases/index.jsx:
```diff
@@ -0,0 +1,54 @@
+			( this.props.siteId === nextProps.siteId ) ) {
+			return;
+		}
+		this.requestSitePurchases( nextProps );
```
What's the reason for using a condition to exit the method whereas we could save one line with:

```javascript
if ( ! nextProps.fetchingSitePurchases && nextProps.siteId &&
  ( this.props.siteId !== nextProps.siteId ) ) {
  this.requestSitePurchases( nextProps );
}
```

Tug replied:
I prefer to write the condition when we're not updating

stephanethomas commented on client/components/data/query-site-purchases/index.jsx:
```diff
@@ -0,0 +1,54 @@
+export default connect(
+	state => {
+		return {
+			fetchingSitePurchases: isFetchingSitePurchases( state )
```
Should we really get rid of the `is` prefix here? We use it abundantly for this sort of flags.

gziolo replied:
`isFetchingSitePurchases` it's selector function, looks fine :)

stephanethomas replied:
Just to clarify, I'm suggesting we don't drop the `is` prefix in the property:

```javascript
isFetchingSitePurchases: isFetchingSitePurchases( state )
```

Tug replied:
Replacing those with `requesting` to be more consistent with other Query Components

stephanethomas commented on client/components/data/query-site-purchases/index.jsx:
```diff
@@ -0,0 +1,54 @@
+class QuerySitePurchases extends Component {
+	constructor( props ) {
+		super( props );
+		this.requestSitePurchases = this.requestSitePurchases.bind( this );
```
Do we really need to bind `requestSitePurchases()` to `this`? As far as I can see it's not used by any event handler.

Tug replied:
Yep you're right, we don't need that 👍

stephanethomas commented on client/components/data/query-site-purchases/README.md:
```diff
@@ -0,0 +1,8 @@
+Query Site Purchases
+================
+
+`<QuerySitePurchases />` is a React component used in managing network requests for sites/purchases (across all users).
```
There is a small typo here: `sites/purchases` should be `site purchases`.

stephanethomas commented on client/components/data/query-user-purchases/index.jsx:
```diff
@@ -0,0 +1,54 @@
+			( this.props.userId === nextProps.userId ) ) {
+			return;
+		}
+		this.requestUserPurchases( nextProps );
```
Same remark than for `QuerySitePurchases.componentWillReceiveProps()`.

stephanethomas commented on client/state/purchases/selectors.js:
```diff
@@ -1,7 +1,20 @@
-import assign from 'lodash/assign';
-import find from 'lodash/find';
+/**
+ * Return user's purchases from state object
```
Are we really only returning user's purchases here? What about site's purchases? It looks like this can be any of them.

Tug replied:
Yep true, they are both mixed into the same array. Updating...

Tug commented:
Thank you all for reviewing, I think we're ready for a final review

drewblaisdell commented:
Thanks for catching that, @stephanethomas. This is actually ready to merge, now. :)

mtias commented:
@Tug thanks for working on this. A pattern we have settled on for query components is that "site" is an argument you can choose to pass. That way we have just a single `QueryPosts` and if you set a `siteId` you'll get what a `QuerySitePosts` would give.

I think this case is suitable for that as well, just one `QueryPurchases` where you can set a `siteId`. What do you think?

Tug commented:
@mtias I see, so we could merge `QuerySitePurchases` and `QueryUserPurchases` into a single component and have the following behavior:
- no params provided: won't fetch anything
- `siteId`: will fetch site purchases
- `userId`: will fetch user purchases
- `siteId` and `userId`: will fetch both site and user purchases

Is this what you are proposing here?

mtias commented:
I think that makes sense, yes. Is it possible to fetch purchases across all sites if the user can see them? Because that's what no params usually means for us.

Tug commented:
I don't think this is ever needed in any components at the moment.
From what I understood purchases are needed at 2 places. The first one is where we manage your own purchases (at `/purchases`). And the other one is in Site Settings to check if a site can be deleted (it can't if it has an active subscription). I believe listing purchases on all sites as no applications at the moment.

mtias commented:
Then perhaps we can assume `QueryPurchases` has `userId` set to current user?

Tug commented:
Yep could work! The only issue I have with this one is when we load `Site Settings` we're going to make an unnecessary request to retrieve the list of purchases for the current user even though we don't need it yet.

mtias commented:
Why do you think so? In site-settings I assume we'd be doing `QueryPurchases siteId={ currentSiteId }` and just `QueryPurchases` in "Me" section.

Tug commented:
Oh ok makes sense!
So in the end, let's just forget about the multisite version, this is what our new component would do:
- no params provided: equivalent to `userId={ currentUser.ID }`
- `userId`: will fetch user purchases
- `siteId`: will fetch site purchases
- `siteId` and `userId`: will fetch both site and user purchases

mtias commented:
That looks great. Just that the last one would be "will fetch user purchases for the given site", no?

drewblaisdell commented:
@mtias The site-specific purchases endpoint fetches all purchases for a site, regardless of the user that made the purchase, in order to determine if a site has active upgrades and shouldn't be deleted.

mtias commented:
@drewblaisdell ok, then we probably don't want the last case at all. If you set a `siteId`, `userId` is meaningless?

drewblaisdell commented:
@mtias Yep, that is correct. Consolidating these components into one that handles the first three cases looks like an improvement though. :+1:
